kanren/goals.py

```diff
@@ -34,42 +27,38 @@ def tailo(tail, coll):
         heado
         conso
     """
-    if isinstance(coll, (tuple, list)):
-        return (fail if len(coll) == 0 else (eq, tail, coll[1:]))
-    else:
-        head = var()
-        return (eq, LCons(head, tail), coll)
+    return (eq, cons(var(), tail), coll)
 
 
-def conso(h, t, l, base_type=tuple):
-    """ Logical cons -- l[0], l[1:] == h, t
+def conso(h, t, l):
+    """ cons h + t == l
+    """
+    return (eq, cons(h, t), l)
 
-    Parameters
-    ==========
-    base_type: type
-        The empty collection type to use when all terms are logic variables.
+
+def nullo(l):
+    """A relation asserting that its argument is a "Lisp-like" null.
+
+    For un-unified logic variables, it unifies with `None`.
+
+    See `is_null`.
     """
-    if isinstance(l, (tuple, list)):
-        if len(l) == 0:
-            return fail
-        else:
-            return (conde, [(eq, h, l[0]), (eq, t, l[1:])])
-    elif isinstance(t, tuple):
-        return eq(type(t)((h,)) + t, l)
-    elif isinstance(t, list):
-        return eq(type(t)([h]) + t, l)
-    else:
-        return (
-            lall,
+    def _nullo(s):
+        _s = reify(l, s)
+        if isvar(_s):
+            yield unify(_s, None, s)
+        elif is_null(_s):
+            yield s
 
-            # The definition of conso. This means that l can be unified with
-            # an LCons object (head + tail).
-            (eq, LCons(h, t), l),
+    return _nullo
 
-            # A "type declaration" for the tail. This means that the first goal
-            # found will simplify to a list with no extra unbound variables.
-            (lany, (eq, t, base_type()), (eq, t, LCons(var(), var())))
-        )
+
+def itero(l):
```

Comment:
Could you add a docstring describing what this does?

Reply:
Added.